### PR TITLE
Russian revolver

### PR DESCRIPTION
**Pull request by MACMAN2003**
## About the PR

Adds in a joke revolver "gun" that is incapable of being used in combat, but is instead used to play Russian Roulette. It can only be used by the player holding it, and is only capable of harming the player holding it.

## Why / Balance

It'd make for some hilarious additions to the spess guy funny moments compilations. As for balance, I'm pretty sure it'd just add a few more patients to medbay each round.

## Technical details

I don't know if I can explain the technical details because I barely understand them myself lol.

## Media

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

https://github.com/space-wizards/space-station-14/assets/20659419/3b96960a-ae10-4643-a616-5db3e483bf1e

## Breaking changes

Content.Shared.Weapons.Ranged.Systems.SharedGunSystem.Revolver.Cycle() was changed from private to public, and a new method "RouletteifyRevolver" was added into it.

**Changelog**

:cl: MACMAN2003, forthbridge, Ubaser
- add: Added a revolver to the Barkeep's booze locker that can be used to play Russian Roulette.

### CODE REVIEW

**Comment by lzk228** [comment]
#8524 :trollface:

**Comment by Equivocateur** [comment]
> #8524 :trollface:

Not a real gun per se, and it has unique functionality.

**Comment by Errant-4** [comment]
It's kind of weird conceptually to introduce a gun that looks like a gun, but arbitrarily can only be used on yourself

**Comment by deltanedas** [comment]
it should be possible to use any gun on yourself, in the ideal world

**Comment by Equivocateur** [comment]
> it should be possible to use any gun on yourself, in the ideal world

/suicide with gun in hand if gun has a bullet in chamber

**Comment by BRINGit34** [comment]
Maybe it should be like one of those cartoon guns where the barrel is pointing the wrong way. Just to make it more clear

**Comment by BolloTea** [comment]
Finally some normal bar entertainment 🤡

**Comment by BolloTea** [comment]
All that remains is to make a reverse pistol for the antagonist clown

**Comment by Partmedia** [comment]
It would be better if this were not a separate weapon, but rather a change that allowed the normal revolver to target yourself. I also don't like the culturally insensitive use of "Russian revolver" and "Russianize", because this sounds like an insult to a whole cultural group.

**Comment by BolloTea** [comment]
> It would be better if this were not a separate weapon, but rather a change that allowed the normal revolver to target yourself. I also don't like the culturally insensitive use of "Russian revolver" and "Russianize", because this sounds like an insult to a whole cultural group.

Excuse me, but do you have any ideas on what to call a revolver for playing "Russian Roulette"?  Besides, there is nothing offensive in this.

**Comment by MACMAN2003** [comment]
> It would be better if this were not a separate weapon, but rather a change that allowed the normal revolver to target yourself. I also don't like the culturally insensitive use of "Russian revolver" and "Russianize", because this sounds like an insult to a whole cultural group.
Do you have any suggestions for alternative names for the gun and/or the "Russianize" method?

**Comment by LankLTE** [comment]
Don't really see how the names are culturally insensitive (any more than Russian Roulette itself would be, which to my knowledge isn't). That said I do agree that having this as its' own revolver seems kind of weird, especially when the revolver looks like a normal gun instead of a cartoony one with the barrel turned around (which would at least warrant looking funny)

**Comment by LankLTE** [comment]
I forgot that this was for the bar- that'd obviously make it difficult to implement as a normal revolver without giving the barkeep their own revolver, which would cause other issues. It's still preferred that this isn't its' own gun, so some other solution would be appreciated if you could find one.
Also, the method name should be changed. Something like MakeRoulette could work better

**Comment by BolloTea** [comment]
The cartoon version with the barrel upside down is not appropriate.  Yes, it wouldn't be bad if every weapon could have a suicide function.  But a separate revolver for playing roulette is the best solution not to be a full-fledged weapon.  This is just a way to test your nerves of steel with friends, after a friendly drinking session in a bar, or as an emergency way to exit the game.

**Comment by Partmedia** [comment]
Maybe you could investigate adding a component field to the `GunComponent` that would allow the shooter to shoot themselves, setting this to false by default.

**Comment by forgotmyotheraccount** [comment]
> > it should be possible to use any gun on yourself, in the ideal world
> 
> /suicide with gun in hand if gun has a bullet in chamber

This is the clear answer. Introducing a new weapon called the "russian revolver" that can only be used on yourself seems silly, no matter what angle you look at it from.

Ideally someone who inputs /suicide while also holding a firearm should have the trigger pulled, and if it's got a lethal round in the chamber it should discharge and kill. If not, it fails. There should be other contextual /suicide methods, like slitting your throat with blades. As it stands /suicide is just a ragequit when it could be a legitimate RP option.

**Comment by Hmeister-fake** [comment]
My one tidbit is the ammo, cause if it comes with real 45. ammo then a fella with an actual revolver can use it but for real. Maybe make it kill you the same way with any ammo type and fill it with practice rounds?

**Comment by AJCM-git** [comment]
I think adding the functionality to "shuffle" the chamber/ammo of any gun with a verb, and then allowing clicking on yourself (with a do_after of course), /suicide, or both to kill you if there is a bullet in the chamber would make a decent implementation for russian roullete.

**Comment by forgotmyotheraccount** [comment]
> My one tidbit is the ammo, cause if it comes with real 45. ammo then a fella with an actual revolver can use it but for real. Maybe make it kill you the same way with any ammo type and fill it with practice rounds?

Even blanks can kill you IRL. Practice rounds could put you into crit rather than killing.

**Comment by BolloTea** [comment]
> > > it should be possible to use any gun on yourself, in the ideal world
> > 
> > 
> > /suicide with gun in hand if gun has a bullet in chamber
> 
> This is the clear answer. Introducing a new weapon called the "russian revolver" that can only be used on yourself seems silly, no matter what angle you look at it from.
> 
I bet that if you indicate in the description that this revolver is intended exclusively for playing roulette, then many questions and misunderstandings will go away by themselves.  But regarding suicide, you can do things differently.  Give a short delay to the rest of the weapon before shooting yourself, so that you can cancel the action if it is accidental or just collect your thoughts.  And the roulette revolver itself has the possibility of instant suicide, because you have already consciously taken a step towards this game.

**Comment by BolloTea** [comment]
Simply, the essence of this pr is not only to add suicide, but also to add an inferior weapon to the bartender, aimed at the entertainment of the station crew)

**Comment by BolloTea** [comment]
Хочу отметить что ЛЕГАЛЬНОГО, а для того чтобы игроки лучше запомнили его, то для этого стоит добавить отдельную текстуру

**Comment by UbaserB** [comment]
Hello, i’m the one who designed the idea of this concept based on ss13’s russian revolvers. As it stands, it is not meant to be a form of suicide as despite the fact it can kill you because /suicide currently works far too well for that. The reason why there is this specific revolver is primarily because from my experience playing russian roulette is way too funny and if it doesn’t right out kill you there isn’t that comedy factor in play. I don’t mind changing changing a few things to specify that its meant to be for the game, and i dont believe the bullet is an issue either since the only other revolvers in the game are the detective’s, syndicate’s and the deckard (if i remember correctly). About another point, i understand that “accidentally killing yourself” is an issue as its an instakill; and i was thinking that requiring harm mode to be active could be a solution for that. We just didnt know how to go about coding it and thought it wasnt exactly necessary so we left it out.

**Comment by UbaserB** [comment]
> Simply, the essence of this pr is not only to add suicide, but also to add an inferior weapon to the bartender, aimed at the entertainment of the station crew)
> I want to note that it is LEGAL, and in order for players to remember it better, it is worth adding a separate texture for this!
> I'm done, sorry.

i copied the inspector sprites, did a rehue and changed some minor details since i thought it would fit better into the game’s style. Should i resprite it entirely then?

**Comment by UbaserB** [comment]
> I forgot that this was for the bar- that'd obviously make it difficult to implement as a normal revolver without giving the barkeep their own revolver, which would cause other issues. It's still preferred that this isn't its' own gun, so some other solution would be appreciated if you could find one.
> Also, the method name should be changed. Something like MakeRoulette could work better

Just clearing it up, its not technically a gun. Its a joke item, similar to soap or space glue. It simply looks like a revolver and (partially) acts like one.

**Comment by asperger-sind** [comment]
> Just clearing it up, its not technically a gun. Its a joke item, similar to soap or space glue. It simply looks like a revolver and (partially) acts like one.

Soap and glue both have legitimate gameplay appliances, this PR adds an item that has exclusively one purely esoteric feature.

**Comment by UbaserB** [comment]
> > Just clearing it up, its not technically a gun. Its a joke item, similar to soap or space glue. It simply looks like a revolver and (partially) acts like one.
> 
> Soap and glue both have legitimate gameplay appliances, this PR adds an item that has exclusively one purely esoteric feature.

soap can only slip people, in 13 they can be used to clean but that purpose currently isnt in 14. Glue, maybe in one crafting recipe like i think improv shotgun uses it surre, but most of the time clown uses it for annoying people by applying it onto objects and it has no other purpose.

**Comment by RedNiloc** [comment]
I think having something like this in the game would be fun. I think it should be a syndi items, or just a clown item.

**Comment by asperger-sind** [comment]
My 2c on the matter (devi asked for takes on the pr in the discord)

The sprite should be a normal revolver with a curved barrel facing towards the user, the name should also be changed to "Russian Roulette Revolver" rather than just staying as "Russian Revolver" because that (along with the current sprites) can be interpreted as a "normal" *Russian* revolver, also change the method name to "make roulette" as Lank suggested.

Ideally all guns would be able to shoot the user but that's probably out of scope, even if that gets coded the barkeep still could have a Russian roulette revolver because its the big silly

Overall good PR, I don't see a reason to not have this in-game.

**Comment by MaloTV** [comment]
Finally an easy way to kill myself (that fucking mouse AHHHHHHHH)

**Comment by Maksim22848** [comment]
> It would be better if this were not a separate weapon, but rather a change that allowed the normal revolver to target yourself. I also don't like the culturally insensitive use of "Russian revolver" and "Russianize", because this sounds like an insult to a whole cultural group.

Нас такое никак не оскорбляет, мы наоборот сами посмеёмся над этим лол

**Review by ficcialfaint — CHANGES_REQUESTED** [review]

File: Resources/Prototypes/Entities/Objects/Fun/russian_revolver.yml [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+      tags:
+        - CartridgeMagnum
+    proto: CartridgeMagnum
+    capacity: 6
+    chambers: [ True, True, True, True, True, True ]
+    ammoSlots: [ null, null, null, null, null, null ]
+    soundEject:
+      path: /Audio/Weapons/Guns/MagOut/revolver_magout.ogg
+    soundInsert:
+      path: /Audio/Weapons/Guns/MagIn/revolver_magin.ogg
```
Comment (ficcialfaint) [inline Resources/Prototypes/Entities/Objects/Fun/russian_revolver.yml:33]: you don't need to write the default values in the prototype
```suggestion
    chambers: [ True, True, True, True, True, True ]
    ammoSlots: [ null, null, null, null, null, null ]
```

**Review by ficcialfaint — CHANGES_REQUESTED** [review]

File: Content.Server/RussianRevolver/RussianRevolverComponent.cs [hunk 2]
```diff
@@ -0,0 +1,18 @@
+using Content.Shared.Damage;
+using Robust.Shared.Audio;
+
+namespace Content.Server.RussianRevolver;
+
+[RegisterComponent]
+public sealed partial class RussianRevolverComponent : Component
```
[the remark this answers is not in the document]
Comment (MACMAN2003) [reply Content.Server/RussianRevolver/RussianRevolverComponent.cs:7]: what does an access attribute do

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
This should just be an action to shoot a gun at yourself as we already have spins. If you really want add a separate action that runs the suicide command rather than just shooting yourself.

File: Content.Server/RussianRevolver/RussianRevolverComponent.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (MACMAN2003) [reply Content.Server/RussianRevolver/RussianRevolverComponent.cs]: i added one but i'm not sure i added it correctly

**Comment by UbaserB** [comment]
> This should just be an action to shoot a gun at yourself as we already have spins. If you really want add a separate action that runs the suicide command rather than just shooting yourself.

by action, do you mean the left sidebar thing?

**Comment by UbaserB** [comment]
nevermind i think i know

**Comment by MACMAN2003** [comment]
found a bug, the bullets that are inserted into the revolver don't get set as shot

**Comment by BolloTea** [comment]
The curved trunk is certainly comical, but it would be more suitable for a clown and not a bartender.  So, a curved trunk is a big stupidity that should not be added.  I think many will agree with this

**Comment by BolloTea** [comment]
>About another point, i understand that “accidentally killing yourself” is an issue as its an instakill; and i was thinking that requiring harm mode to be active could be a solution for that. We just didnt know how to go about coding it and thought it wasnt exactly necessary so we left it out.

Sorry, translation difficulties.  I meant the delay for other weapons, and not specifically for the Russian revolver.  Sorry if I misled you.  Nevertheless, your pr is what many have been waiting for, especially old like me.  So, thank you very much for your work.
Удачи🤠

**Comment by BolloTea** [comment]
> My 2c on the matter (devi asked for takes on the pr in the discord)
> 
> The sprite should be a normal revolver with a curved barrel facing towards the user, the name should also be changed to "Russian Roulette Revolver" rather than just staying as "Russian Revolver" because that (along with the current sprites) can be interpreted as a "normal" _Russian_ revolver, also change the method name to "make roulette" as Lank suggested.
> 
> Ideally all guns would be able to shoot the user but that's probably out of scope, even if that gets coded the barkeep still could have a Russian roulette revolver because its the big silly
> 
> Overall good PR, I don't see a reason to not have this in-game.

And to avoid confusion, you need to make a correct description, and not change the name

File: Content.Server/RussianRevolver/RussianRevolverComponent.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (MACMAN2003) [reply Content.Server/RussianRevolver/RussianRevolverComponent.cs]: i figured out what i was doing wrong and fixed it

**Comment by UbaserB** [comment]
> > My 2c on the matter (devi asked for takes on the pr in the discord)
> > The sprite should be a normal revolver with a curved barrel facing towards the user, the name should also be changed to "Russian Roulette Revolver" rather than just staying as "Russian Revolver" because that (along with the current sprites) can be interpreted as a "normal" _Russian_ revolver, also change the method name to "make roulette" as Lank suggested.
> > Ideally all guns would be able to shoot the user but that's probably out of scope, even if that gets coded the barkeep still could have a Russian roulette revolver because its the big silly
> > Overall good PR, I don't see a reason to not have this in-game.
> 
> And to avoid confusion, you need to make a correct description, and not change the name

we also did this

**Comment by VektorZ1** [comment]
Would be many times cooler to add the ability to point any revolver at yourself for a game of Russian roulette.

**Comment by Emisse** [comment]
you should just be able to shoot yourself with a gun

**Comment by deltanedas** [comment]
actions are bloat and would be annoying, imo it should be when you click yourself with a gun it makes a popup saying you place the barrel in your mouth and close your eyes, then when its up inflict 2x or whatever damage because its going right into your brain

**Comment by MACMAN2003** [comment]
i love waking up to a merge conflict
how do i fix that

**Comment by UbaserB** [comment]
> you should just be able to shoot yourself with a gun

An issue with letting this happen means that the average person wouldnt be able to play russian revolver. You’re not allowed to carry a gun on your person unless you are security or the bartender, and this is simply meant to be a friendly game of russian roulette. However, i suppose it wouldn’t be hard to add an additional feature of using /suicide command while holding a gun to kill yourself with certainty.

**Comment by UbaserB** [comment]
> This should just be an action to shoot a gun at yourself as we already have spins. If you really want add a separate action that runs the suicide command rather than just shooting yourself.

After trying to figure out how to implement this, we came to the realisation that not only should this change be done in a separate PR, but it is way out of scope for what we are trying to implement. Currently as it stands though, there is no issue with the current code and functions normally ingame.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
From discord:
no new guns
shouldn't "roulettify"
shouldn't be its own component
should just use an action
it also shouldn't be a 98.5% chance of spawning inside the booze locker

**Comment by MACMAN2003** [comment]
it's not a real gun, it's a toy.
as for switching it to an action, i have no idea how to implement that.

**Comment by Partmedia** [comment]
I'm going to go ahead and close this, because to address the reviewer comments this would require a bit of a rework. Please feel free to open a new PR for this once it is ready.

**Pull request closed without merging**